cpp/src/arrow/compute/kernels/vector_selection_filter_internal.cc

```diff
@@ -998,6 +998,11 @@ class FilterMetaFunction : public MetaFunction {
   Result<Datum> ExecuteImpl(const std::vector<Datum>& args,
                             const FunctionOptions* options,
                             ExecContext* ctx) const override {
+    if ((args[1].kind() != Datum::ARRAY) &&
+        (args[1].kind() != Datum::CHUNKED_ARRAY)) {
+        return Status::NotImplemented("Filter should be array-like");
```

Comment:
Can we make this `TypeError`?

Reply:
Sounds good to me. I got this from [L945](https://github.com/apache/arrow/blob/ae16bde4b36fe5a404c8ca22fab769db0d652543/cpp/src/arrow/compute/kernels/vector_selection_filter_internal.cc#L945). Should change that one as well?

Reply:
Yes, definitely.

Reply:
Okay, changing and updating the tests in pyarrow that need updating (I don't know if this will affect downstream tests for other subprojects)

Reply:
Looks like it didn't affect anything on the pyarrow/pytest side of things. Changed!